source/codegen/metadata_validation.py

```diff
@@ -0,0 +1,287 @@
+from typing import Any, Dict, Iterable, List, Set
+from schema import Schema, And, Optional, Or, Use
+import common_helpers
+
+
+# Rules that can be suppressed
+class RULES:
+    # In general, arrays should have sizes passed in to the underlying C API.
+    ARRAY_PARAMETER_NEEDS_SIZE = "ARRAY_PARAMETER_NEEDS_SIZE"
+    # The gRPC layer can determine the size of input arrays, so we should not require
+    # callers to pass in the size. We should not add any new suppressions of this type.
+    INPUT_ARRAY_SHOULD_NOT_HAVE_PASSED_IN_SIZE = "INPUT_ARRAY_SHOULD_NOT_HAVE_PASSED_IN_SIZE"
+    # In general, enums shouldn't have duplicate values. This is helpful for catching typos
+    # although there are a few enums that have legitimate duplicates.
+    ENUMS_SHOULD_NOT_HAVE_DUPLICATE_VALUES = "ENUMS_SHOULD_NOT_HAVE_DUPLICATE_VALUES"
+
+
+def validate_metadata(metadata: dict):
+    try:
+        for function_name in metadata['functions']:
+            validate_function(function_name, metadata)
+        for attribute_group in common_helpers.get_attribute_groups(metadata):
+            for attribute_id in attribute_group.attributes:
+                validate_attribute(attribute_group.attributes[attribute_id], metadata)
+        function_enums = get_function_enums(metadata['functions'])
+        attribute_enums = get_attribute_enums(metadata)
+        used_enums = function_enums.union(attribute_enums)
+        for enum_name in metadata['enums']:
+            validate_enum(enum_name, used_enums, metadata)
+    except Exception as e:
+        raise Exception(
+            f"Failed to validate {metadata['config']['namespace_component']}") from e
+
+DOCUMENTATION_SCHEMA = Schema(
+    {
+        "description": str,
+        Optional("note"): str,
+        Optional("table_body"): list,
+        Optional("caution"): str,
+    }
+)
+
+SIZE_SCHEMA = Schema(
+    {
+        'mechanism': And(str, lambda s: not common_helpers.is_unsupported_size_mechanism_type(s)),
+        'value': Or(str, int),
+        Optional('value_twist'): str,
+        Optional('tags'): [str],
+    },
+)
+
+PARAM_SCHEMA = Schema(
```

Comment:
note: I decided to explicitly list the keys that are valid even if we don't use most of them. My idea was that this will prevent us from accepting typos in keys that we don't want. Sadly the list of "keys we use somewhere" is a bit longer than I expected...